prov/verbs: Improve getinfo attributes check

a-ilango opened the pull request:
* Create a global verbs_info structure which will be used for 
  verifying fi_info structures.
* Group together the calls to get device attributes.
* Replace default values for attributes with actual ones from the device.
* Remove prov_name assignment. It would be done by fi_getinfo.
* Copy attributes from hints to info being returned.

Signed-off-by: Arun C Ilango <arun.ilango@intel.com>

patrickmacarthur commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -557,110 +590,192 @@ static int fi_ibv_rai_to_fi(struct rdma_addrinfo *rai, struct fi_info *fi)
+	init_attr.cap.max_send_wr = 1;
+	init_attr.cap.max_recv_wr = 1;
+	init_attr.cap.max_send_sge = 1;
+	init_attr.cap.max_recv_sge = 1;
```
Any reason we aren't pulling these values from hints->tx_attr and hints->rx_attr?

a-ilango replied:
The qp is created here only to verify if hints->tx_attr->inject_size is accepted by the device. There isn't a way to get  max_inline_data directly. On the other hand, max_send_sge and max_recv_sge can be got from device_attr.max_sge. Hence there is no need to pass along hints->tx_attr->iov_limit here.

shefty replied:
Consider moving this into a function, so that the intent of the code is clear.

a-ilango replied:
Good idea. I will move this inside a function.

patrickmacarthur commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -363,112 +391,151 @@ static int fi_ibv_check_ep_attr(const struct fi_ep_attr *attr)
+		FI_INFO(&fi_ibv_prov, FI_LOG_CORE,
+			"Given tx_attr->size is greater than supported\n");
+		return -FI_ENODATA;
+	}
```
Shouldn't this be checking against `rx_attr->size`?

a-ilango replied:
Thanks! Will correct that.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -557,110 +590,192 @@ static int fi_ibv_rai_to_fi(struct rdma_addrinfo *rai, struct fi_info *fi)
+	}
+
+	info->ep_attr->max_msg_size 		= port_attr.max_msg_sz;
+	info->ep_attr->max_order_raw_size 	= port_attr.max_msg_sz;
```
Verb devices do not support RAW ordering.  This value should be 0.  The original code (removed above) was incorrect as well.

a-ilango replied:
But Infiniband spec (table 76) says RDMA Read after RDMA Write will have order maintained.

shefty replied:
You're right.  I was thinking of write after read.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -752,31 +861,30 @@ static int fi_ibv_getinfo(uint32_t version, const char *node, const char *servic
+		ret = fi_ibv_create_info(hints, &verbs_info);
+		if (ret)
+			return ret;
+	}
```
This isn't thread safe.  At least add a comment to handle multiple threads calling fi_getinfo simultaneously.

a-ilango replied:
Thanks! I'll add a comment for now and later fix it to be thread safe.

patrickmacarthur commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -557,110 +590,192 @@ static int fi_ibv_rai_to_fi(struct rdma_addrinfo *rai, struct fi_info *fi)
+	info->domain_attr->max_ep_rx_ctx 	= device_attr.max_qp;
+	info->tx_attr->iov_limit 		= device_attr.max_sge;
+	info->tx_attr->rma_iov_limit 		= device_attr.max_sge;
+	info->rx_attr->iov_limit 		= device_attr.max_sge;
```
We should probably also set `tx_attr->size` and `rx_attr->size` based on device/queue pair attributes (`device_attr.max_qp_wr`?), rather than hardcoding a value.

shefty replied:
> We should probably also set tx_attr->size and rx_attr->size based on
> device/queue pair attributes (device_attr.max_qp_wr?), rather than
> hardcoding a value.

This would result in the user *always* allocating the maximum size QP, which is highly undesirable.  The intent is to provide an 'optimal' value.  However, only the vendors can really provide that.  Selecting a hard-code value that attempts to bridge the gap between reasonable performance with minimal memory footprint seems ideal.  There's really no reason to construct QPs with queues 65,000 entries deep.

patrickmacarthur replied:
Ok, that makes sense.

a-ilango commented:
Thanks for the review. I have updated the PR with suggested changes. Also added thread safety for verbs_info initialization, renamed fi_ibv_create_info and modified its signature. Please review.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -224,16 +224,15 @@ static int fi_ibv_sockaddr_len(struct sockaddr *addr)
 	}
 
-	if (attr->prov_version > VERBS_PROV_VERS) {
+	if (attr->prov_version > info->fabric_attr->prov_version) {
```
Because this is checking the value from a parameter, the check looks wrong.  The check must be against the version defined by the verbs code, and not through some parameter, which could have come from the app.

I *think* the confusion is that the 'info' parameter is really referring to the global variable verbs_info.  If that's the case, then I would reference verbs_info directly, as a global to indicate the intent.

shefty commented:
This now conflicts with other updates.  Merging manually.
